org.eclipse.lsp4e/src/org/eclipse/lsp4e/DocumentContentSynchronizer.java

```diff
@@ -163,6 +169,42 @@ private boolean createChangeEvent(DocumentEvent event) {
 		return true;
 	}
 
+	private boolean serverSupportsWillSaveWaitUntil() {
+		ServerCapabilities serverCapabilities = languageServerWrapper.getServerCapabilities();
+		if(serverCapabilities != null ) {
+			Either<TextDocumentSyncKind, TextDocumentSyncOptions> textDocumentSync = serverCapabilities.getTextDocumentSync();
+			if(textDocumentSync.isRight()) {
+				TextDocumentSyncOptions saveOptions = textDocumentSync.getRight();
+				return saveOptions != null && saveOptions.getWillSaveWaitUntil();
+			}
+		}
+		return false;
+	}
+
+	public void documentAboutToBeSaved() {
+		if (!serverSupportsWillSaveWaitUntil()) {
+			return;
+		}
+
+		TextDocumentIdentifier identifier = new TextDocumentIdentifier(fileUri.toString());
+		WillSaveTextDocumentParams params = new WillSaveTextDocumentParams(identifier, TextDocumentSaveReason.Manual); // actually unknown at this point in time
+		CompletableFuture<List<TextEdit>> willSaveUntilFuture = languageServerWrapper.getInitializedServer().thenApplyAsync(ls -> {
+			try {
+				return ls.getTextDocumentService().willSaveWaitUntil(params).get(500, TimeUnit.MILLISECONDS);
+			} catch (InterruptedException | ExecutionException e) {
+				LanguageServerPlugin.logError(e);
+				return Collections.emptyList();
+			} catch (TimeoutException e) {
+				return Collections.emptyList();
+			}
+		});
+		try {
+			willSaveUntilFuture.thenAccept(edits -> LSPEclipseUtils.applyEdits(document, edits)).get(500, TimeUnit.MILLISECONDS);
```

Comment:
I think it reads and overall performs better if you run
```
languageServerWrapper.getInitializedServer().thenComposeAsync(ls -> ls.getTextDocumentService().willSaveWaitUntil(params).thanApplyAsync(edits -> LSPEclipseUtils.applyEdits(document, params))
```
But this is indeed prone to race conditions (eg `save` may be invoked before `willSave...` is queried. So the willSave future should probably be stored on the listener and the actual `save` should take care of first waiting for the willSave future if necessary.

Reply:
By looking at `org.eclipse.core.internal.filebuffers.FileStoreFileBuffer.commit(IProgressMonitor, boolean)`, the method will fire always first `fireStateChanging`, which iterates through the listeners and calls the handlers. As far as I understand the iterator waits for each handler to finish. Then `fireDirtyStateChanged` is called. So in this code path, `save` will always be called after `willSave...` is finish. Do you have other code paths in mind?

Reply:
ok, got it. It's indeed fine to make the listener blocking then.

Reply:
Please try the change mentioned in earlier comment anyway if possible, it makes code simpler and allows to only join on a single future and thus has a better control on the timeout being a "global" timeout.

Reply:
In addition to my last comment, `save...` should not wait for the willSave future, the work doing in that future must be finished by the time the buffer is saved (otherwise the buffer would be saved, the work would modify the buffer again in the background and the buffer would get dirty). Since we call `DocumentContentSynchronizer.documentSaved` on the DirtyStateChanged event, it is too late to wait on documentSaved.

I will take a look at the comment on how to rewrite it for better readability and performance, I have briefly looked at it, and the code
`languageServerWrapper.getInitializedServer()
		.thenComposeAsync(ls -> ls.getTextDocumentService().willSaveWaitUntil(params))
		.thanApplyAsync(edits -> LSPEclipseUtils.applyEdits(document, edits));`
has the compiler error

The method thanApplyAsync((<no type> edits) -> {}) is undefined for the type CompletableFuture<List<TextEdit>>.
I need to understand that one to be able to submit a newer version, and currently I do not get it yet. I find this API quite difficult to work with :(.

Reply:
> n addition to my last comment, save... should not wait for the willSave future, the work doing in that future must be finished by the time the buffer is saved

Yes, the "applyEdits" would need to happen before.

> The method thanApplyAsync(( edits) -> {}) is undefined for the type CompletableFuture<List>.

Sorry I meant `.thenAcceptAsync()`.

Reply:
thanks, that worked